woocommerce/woocommerce-ios
# [Product Multi-Selection] Decouple clear selection from toggle variations

Feel free to leave this review to post-HACK week 👍

Closes: #9043 

## Description
This PR decouples the "Clear Selection" logic from using "Toggle Variations" as a flag, setting its own property in the view model making the difference between both explicit. We want two different behaviours for Products, and Coupons:
- Products: Has no "Clear Selection" button. Product Variations cannot be toggled (cannot select all variations at once)
- Coupons: Has "Clear Selection" button. Variations can be toggled (can select all variations at once)

## Changes
- Adds a `isClearSelectionEnabled` property to `ProductSelectorViewModel`. This is `true` by default unless set otherwise (our case).
- Adds `isProductMultiSelectionM1Enabled` to `MockFeatureFlagService`
- Adds Unit Tests.

## Testing instructions
On a Store with Product Variations:
1. Switch `productMultiSelectionM1` feature flag to `true`
2. Go to Orders > Tap on `+` > Tap on `+ Add Product`
3. Select a Product > See how the "Clear Selection" button does not appear at the top of the screen
4. Tap on the `o` checkbox for the Product that represents all variations > See how cannot be toggled.

5. Go to Menu > Coupons (enable the feature if is disabled in Experimental Features) > Tap on `+` > Tap on Percentage discount > Tap on All Products > Confirm the "Clear Selection" button still appears at the top of the screen.
6. Tap on the `o` checkbox for the Product that represents all variations > See how is toggled and selects all existing variations.

## Screenshots

| Coupons: Has "Clear Selection" button. Variations can be toggled  | Order Creation: No "Clear Selection" button. Variations cannot be toggled |
|---|---|
| <img width=450 src="https://user-images.githubusercontent.com/3812076/223731497-e669e5be-a5b0-4a5d-9de2-682dfdec5642.png">  |  <img width=450 src="https://user-images.githubusercontent.com/3812076/223731530-fb443de7-31ae-4f03-ad09-fc08321cb2b1.png"> |



---
